x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java

```diff
@@ -0,0 +1,154 @@
+                .getHits();
+        } catch (SearchPhaseExecutionException e) {
+            // NOTE: searching the target downsample index for the latest tsid failed. We just start the task from scratch.
+            lastRollupTsidHits = new SearchHit[0];
+        }
+        final RollupShardPersistentTaskState initialState = lastRollupTsidHits.length == 0
+            ? new RollupShardPersistentTaskState(RollupShardIndexerStatus.INITIALIZED, null)
+            : new RollupShardPersistentTaskState(
+                RollupShardIndexerStatus.STARTED,
+                Arrays.stream(lastRollupTsidHits).findFirst().get().field("_tsid").getValue()
+            );
+        final RollupShardIndexer rollupShardIndexer = new RollupShardIndexer(
+            (RollupShardTask) task,
+            client,
+            getIndexService(indicesService, params),
+            params.shardId(),
+            params.rollupIndex(),
+            params.downsampleConfig(),
+            params.metrics(),
+            params.labels(),
+            initialState
+        );
+        try {
+            rollupShardIndexer.execute();
+        } catch (IOException e) {
+            throw new ElasticsearchException("Unable to start downsampling shard task [" + task.getPersistentTaskId() + "]", e);
+        }
+    }
+
+    private static IndexService getIndexService(final IndicesService indicesService, final RollupShardTaskParams params) {
+        return indicesService.indexService(params.shardId().getIndex());
+    }
+
+    @Override
+    protected AllocatedPersistentTask createTask(
+        long id,
+        final String type,
+        final String action,
+        final TaskId parentTaskId,
+        final PersistentTasksCustomMetadata.PersistentTask<RollupShardTaskParams> taskInProgress,
+        final Map<String, String> headers
+    ) {
+        final RollupShardTaskParams params = taskInProgress.getParams();
+        return new RollupShardTask(
+            id,
+            type,
+            action,
+            parentTaskId,
+            params.rollupIndex(),
+            params.indexStartTimeMillis(),
+            params.indexEndTimeMillis(),
+            params.downsampleConfig(),
+            headers,
+            params.shardId(),
+            new RollupShardPersistentTaskState(RollupShardIndexerStatus.INITIALIZED, null)
+        );
+    }
+
+    @Override
+    public PersistentTasksCustomMetadata.Assignment getAssignment(
+        final RollupShardTaskParams params,
+        final Collection<DiscoveryNode> candidateNodes,
+        final ClusterState clusterState
+    ) {
+        // NOTE: downsampling works by running a task per each shard of the source index.
+        // Here we make sure we assign the task to the actual node holding the shard identified by
+        // the downsampling task shard id.
+        final ShardId shardId = params.shardId();
+        final ShardRouting shardRouting = clusterState.routingTable().shardRoutingTable(shardId).primaryShard();
```

Comment:
I think any shard is ok for downsampling? Also replica shards.

Reply:
Well in the transport action we set replicas to 0 and we set the original value back for the index only after downsampling is complete.  Here I used only primary shards because I wanted to make sure we are not using some replica that maybe is created before we set the number of replicas to 0.